rfcs/20190116-embedding-partitioned-variable.md

```diff
@@ -0,0 +1,498 @@
+
+However, embeddings are not supported by multi-worker `CollectiveAllReduceStrategy` unless we implement multi-worker allgather primitive in collective ops or always cast their updates to dense tensors and use allreduce to aggregate gradients.
+
+Note: our design proposal below assumes there is an efficient implementation of allgather but any detail of allgather is out of this doc's scope. We'll have allgather implementation in collective ops soon.
+
+
+### Partitioned Layer
+
+
+A layer of a network can sometimes be partitioned. We call it partitioned layer in this document. This is another use case of `PartitionedVariable`.
+
+The main reason people partition a layer is to balance parameter servers' load.
+
+Some layers can be huge. For example the logits layer can be extremely large when there are large number of classes. People may want to shard their computations accordingly to avoid concatenating them, e.g. the implementation of `tf.nn.nce_loss`.
+
+Even for mid-sized layer, people want to partition them to have more even distribution of parameter servers' load because people may get bad luck when variables are assigned to parameter servers in a round-robin fashion. See this [issue](https://github.com/tensorflow/tensorflow/issues/24953).
+
+Partitioned layer is only supported by lower level APIs with or without `ParameterServerStrategy` in 1.x.
+
+
+### Current Usage of PartitionedVariable
+
+
+With TF 1.x API users can create a `PartitionedVariable` via `tf.get_variable` API:
+
+```python
+partitioned_var = tf.get_variable(
+    "embedding",
+    shape=(10000, 128),
+    dtype=tf.float32,
+    partitioner=tf.min_max_variable_partitioner(
+        max_partitions=num_ps_replicas, axis=0))
+```
+
+The `partitioned_var` here is an instance of `PartitionedVariable`.
+
+Alternatively users can use `tf.variable_scope` to apply `partitioner` to all variables created under its scope:
+
+```python
+with tf.variable_scope(partitioner=tf.min_max_variable_partitioner(
+        max_partitions=num_ps_replicas, axis=0)):
+  tf.keras.layers.Embedding(...)
+```
+
+Both these cases will also work when they are under a `ParameterServerStrategy`'s scope. However `MirroredStrategy` and `CollectiveAllReduceStrategy` will [error out](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/distribute/distribute_lib.py#L872) if a `PartitionedVariable` is about to be created.
+
+These APIs `tf.get_variable` and `tf.variable_scope` will be gone in TF 2.0.
+
+
+### Possible Problems with The `PartitionedVariable` Class 
+
+
+Although we have a class called `PartitionedVariable`,  we often concatenate all of their components to pretend it is a single variable.
+
+For example, `tf.matmul(partitioned_var, inputs)` would concatenate the list of variables in the `partitioned_var` before the execution of `matmul` as if the `partitioned_var` is a single variable. With the concatenation, the `partitioned_var` would only serve for loadbalancing purposes.
```

Comment:
And this will in fact be wrong if you want mod partitioning instead of div partitioning.

(another issue is that keeping track of mod vs div is delegated to users who might no longer have access to this information)

Reply:
I was talking about non-embedding loadbalancing variables. I rephrased it and added a paragraph for the ugliness of the `partition_strategy` argument.